cognite_toolkit/cdf_tk/templates.py

```diff
@@ -447,293 +447,281 @@ def build_config(
-    for line in dumped.splitlines():
-        indent = len(line) - len(line.lstrip())
-        if last_indent < indent:
-            if last_variable is None:
-                raise ValueError("Unexpected state of last_variable being None")
-            path = (*path, last_variable)
-        elif last_indent > indent:
-            # Adding some extra space between modules
-            out_lines.append("")
-            indent_reduction_steps = (last_indent - indent) // indent_size
-            path = path[:-indent_reduction_steps]
-
-        variable = line.split(":", maxsplit=1)[0].strip()
-        if local_comment := comments.get((*path, variable)):
-            for line_comment in local_comment["above"]:
-                out_lines.append(f"{' ' * indent}# {line_comment}")
-            if after := local_comment["after"]:
-                line = f"{line} # {after[0]}"
-
-        out_lines.append(line)
-        last_indent = indent
-        last_variable = variable
-    out_lines.append("")
-    return "\n".join(out_lines)
+    @property
+    def key(self) -> str:
+        return ".".join(self.key_path)
 
 
 @dataclass
-class ConfigEntries(UserList):
-    def __init__(self, entries: list[ConfigEntry] | dict | None = None):
-        if isinstance(entries, dict):
-            entries = self._initialize(entries)
+class ConfigYAML(UserDict[tuple[str, ...], ConfigEntry]):
+    """This represents the 'config.yaml' file in the root of the project.
+
+    The motivation for having a specialist data structure and not just a dictionary:
+
+    1. We want to be able to dump the config.yaml file with comments.
+    2. We want to track which variables are added, removed, or unchanged.
+    """
+
+    def __init__(self, entries: dict[tuple[str, ...], ConfigEntry] | None = None):
         super().__init__(entries or [])
-        self._lookup: dict[str, dict[str, ConfigEntry]] = {}
-        for entry in self:
-            self._lookup.setdefault(entry.module, {})[entry.key] = entry
 
-    @staticmethod
-    def _initialize(entries: dict, path: str = "") -> list[ConfigEntry]:
-        results = []
-        if "." in path:
-            path_to, module = path.rsplit(".", maxsplit=1)
-        else:
-            module = path
-            path_to = ""
-        for key, value in entries.items():
-            if isinstance(value, dict):
-                results.extend(ConfigEntries._initialize(value, f"{path}.{key}" if path else key))
-            else:
-                results.append(
-                    ConfigEntry(
-                        key=key,
-                        module=module,
-                        path=path_to,
-                        last_value=value,
-                        current_value=None,
-                    )
+    @classmethod
+    def load(cls, directory: Path | Sequence[Path], existing_config_yaml: str | None = None) -> ConfigYAML:
+        """Loads the config.yaml file from a module directory and all its subdirectories.
```

Comment:
Is this an accurate description of what this method does? From a module directory? Sort of confused me. Looking at the code it looks like you mean default.config.yaml?

Reply:
Improved it now.